packages/aws-cdk/lib/api/garbage-collector.ts

```diff
@@ -0,0 +1,393 @@
+          if (obj.Key) {
+            batch.push(new S3Asset(bucket, obj.Key));
+          }
+        });
+
+        continuationToken = response.NextContinuationToken;
+
+        if (!continuationToken) break; // No more objects to fetch
+      }
+
+      if (batch.length > 0) {
+        yield batch;
+      }
+    } while (continuationToken);
+  }
+
+  /**
+   * Fetches all relevant stack templates from CloudFormation. It ignores the following stacks:
+   * - stacks in DELETE_COMPLETE or DELET_IN_PROGRES stage
+   * - stacks that are using a different bootstrap qualifier
+   *
+   * It fails on the following stacks because we cannot get the template and therefore have an imcomplete
+   * understanding of what assets are being used.
+   * - stacks in REVIEW_IN_PROGRESS stage
+   */
+  private async fetchAllStackTemplates(cfn: CloudFormation, qualifier: string) {
+    const stackNames: string[] = [];
+    await paginateSdkCall(async (nextToken) => {
+      const response = await cfn.listStacks({ NextToken: nextToken }).promise();
+
+      // We cannot operate on REVIEW_IN_PROGRESS stacks because we do not know what the template looks like in this case
+      const reviewInProgressStacks = response.StackSummaries?.filter(s => s.StackStatus === 'REVIEW_IN_PROGRESS') ?? [];
+      if (reviewInProgressStacks.length > 0) {
+        throw new Error(`Stacks in REVIEW_IN_PROGRESS state are not allowed: ${reviewInProgressStacks.map(s => s.StackName).join(', ')}`);
+      }
+
+      // Deleted stacks are ignored
+      stackNames.push(
+        ...(response.StackSummaries ?? [])
+          .filter(s => s.StackStatus !== 'DELETE_COMPLETE' && s.StackStatus !== 'DELETE_IN_PROGRESS')
+          .map(s => s.StackId ?? s.StackName),
+      );
+
+      return response.NextToken;
+    });
+
+    print(chalk.blue(`Parsing through ${stackNames.length} stacks`));
+
+    const templates: string[] = [];
+    for (const stack of stackNames) {
+      let summary;
+      summary = await cfn.getTemplateSummary({
+        StackName: stack,
+      }).promise();
+
+      // Filter out stacks that we KNOW are using a different bootstrap qualifier
+      // This is necessary because a stack under a different bootstrap could coincidentally reference the same hash
+      // and cause a false negative (cause an asset to be preserved when its isolated)
+      const bootstrapVersion = summary?.Parameters?.find((p) => p.ParameterKey === 'BootstrapVersion');
+      const splitBootstrapVersion = bootstrapVersion?.DefaultValue?.split('/');
+      if (splitBootstrapVersion && splitBootstrapVersion.length == 4 && splitBootstrapVersion[2] != qualifier) {
+        // This stack is definitely bootstrapped to a different qualifier so we can safely ignore it
+        continue;
+      } else {
+        const template = await cfn.getTemplate({
+          StackName: stack,
+        }).promise();
+
+        templates.push(template.TemplateBody ?? '' + summary?.Parameters);
```

Comment:
its there, just tacked on since we're doing string lookups anyway

Reply:
Is `Parameters` a string then? Because if not, your final string in the array will look like: `{ ... } [object Object]`

This might indicate a missing test.

Reply:
yup. missing test. fixed now.